lib/puppet/pops/evaluator/closure.rb

```diff
@@ -48,18 +48,39 @@ def call(*args)
   # checks of the argument type/arity validity).
   # @api private
   def invoke(instance, calling_scope, args, &block)
-    call(*args, &block)
+    scopes_to_restore = @enclosing_scope.unset_ephemeral_var(1)
+    begin
+      # Only global scope should be visible to the called function. Other ephemeral scopes must be popped
+      # and later restored
+      call(*args, &block)
+    ensure
+      unless scopes_to_restore.nil?
+        @enclosing_scope.unset_ephemeral_var(1)
```

Comment:
I mentioned this in the scrum. The call_by_name (used when assigning EPP templates) has access to the enclosing scope (as you explained, this is wrong, only inline_epp should have that access). That's why the call_by_name just takes the level and restores it once all parameter defaults have been evaluated.

The invoke is different. It is what differentiates a function call from a lambda call. A function call must only see the global scope. It should not see any ephemeral scopes that are pushed onto it. So all such scopes are temporarily popped during the call and then later restored once the call returns.